README.md
pedigreeTools
=========
The ```pedigreeTools``` package offers a suit of functions for pedigree analyses, 
including: sorting and editing pedigree data, computing inbreeding, 
additive relationships and functions of it (e.g., cholesky derived from a pedigree, 
inverse of a numerator relationship matrix). The package was originally 
co-developed by Douglas M. Bates and Ana I. Vazquez and was incorporated in 
the [lme4](https://CRAN.R-project.org/package=lme4) R-package. Recently, 
pedigreeTools evolved to an stand-alone package. We have added new 
features for pedigree edition, analysis of data from self-pollination 
and for subsetting pedigrees.

Developers: Ana I. Vazquez & Douglas M. Bates.
Paulino Perez-Rodriguez & S. Avadhanam.

### Installing pedigreeTools from GitHub

```R
#R functions related to pedigrees
install.packages(pkg='devtools',repos='https://cran.r-project.org/')  #1# install devtools
library(devtools)                                                     #2# load the library
install_git('https://github.com/Rpedigree/pedigreeTools/')            #3# install pedigreeTools from GitHub
```
### Examples
 
  1. [Completing and sorting a pedigree](https://github.com/Rpedigree/pedigreeTools/blob/master/inst/examples/example1.md).
  2. [Computing inbreeding, additive relationships and funcitons of it](https://github.com/Rpedigree/pedigreeTools/blob/master/inst/examples/example2.md).
  3. [Subsetting a Pedigree using prunePed](https://github.com/Rpedigree/pedigreeTools/blob/master/inst/examples/example3.md).
  4. [Pedigree analyses for self-pollinated species](https://github.com/Rpedigree/pedigreeTools/blob/master/inst/examples/example4.md).
  5. [Cross-validation analyses with pedigreeTools and pedigreemm](https://github.com/Rpedigree/pedigreeTools/blob/master/inst/examples/example5.md).



CHANGELOG
version 0.1 (2018-11-14)
Initial version of the software

version 0.2 (2019-03-06)
Bug fixes, 
The rchk tool (https://github.com/kalibera/rchk) detected 
a memory protection bug in the file pedigree.c (line 46) that 
was fixed in this version. Only three lines of code were modified:

before:

setAttrib(ans, install("F"), allocVector(REALSXP, n));

now has ben replaced with:

SEXP tmp2 = PROTECT(allocVector(REALSXP, n));
setAttrib(ans, install("F"), tmp2);

and a call to the UNPROTECT() macro was added at the end of the function:

UNPROTECT(1);
return ans;


R/test.R
```R

```


R/AllClass.R
#' Pedigree class
#'
#' @export
setClass("pedigree", representation =
	 list(sire = "integer", dam = "integer", label = "character"),
	 validity = function(object) {
	     n <- length(sire <- object@sire)
	     if (length(dam <- object@dam) != n)
		 return("sire and dam slots must be the same length")
	     if (length(object@label) != n)
		 return("'label' slot must have the same length as 'sire' and 'dam'")
	     if(n == 0) return(TRUE)
	     animal <- 1:n
	     snmiss <- !is.na(sire)
	     dnmiss <- !is.na(dam)
	     if (any(sire[snmiss] >= animal[snmiss]) ||
		       any(dam[dnmiss]  >= animal[dnmiss]))
		 return("the sire and dam must precede the offspring")
             if (any(sire[snmiss] < 1 | sire[snmiss] > n) |
                 any(dam[dnmiss] < 1 | dam[dnmiss] > n))
                 return(paste("Non-missing sire or dam must be in [1,",
                              n, "]", sep = ''))
	     TRUE
	 })


src/pedigree.c
#include "pedigree.h"

#ifdef ENABLE_NLS		/** Allow for translation of error messages */
#include <libintl.h>
#define _(String) dgettext ("pedigreemm", String)
#else
#define _(String) (String)
#endif

/* When appropriate, alloca is cleaner than malloc/free.  The storage
 * is freed automatically on return from a function. When using gcc the
 * builtin version is much faster. */
#ifdef __GNUC__
# undef alloca
# define alloca(x) __builtin_alloca((x))
#else
/* this is necessary (and sufficient) for Solaris 10: */
# ifdef __sun
#  include <alloca.h>
# endif
#endif

/** alloca n elements of type t */
#define Alloca(n, t)   (t *) alloca( (size_t) ( (n) * sizeof(t) ) )

/* Functions for the pedigree class */

/**
 * Create the left Cholesky factor of the numerator relationship
 * matrix from a pedigree.
 *
 * @param x a pedigree object
 * @param ans T stored as a non-unit, lower dtCMatrix
 *
 * @return ans with elements modified to incorporate D
 */
SEXP pedigree_chol(SEXP x, SEXP ans)
{
    SEXP Sire = GET_SLOT(x, install("sire"));
    int *ai = INTEGER(GET_SLOT(ans, install("i"))),
	*ap = INTEGER(GET_SLOT(ans, install("p"))),
	*dam = INTEGER(GET_SLOT(x, install("dam"))),
	*sire = INTEGER(Sire), j, n = LENGTH(Sire);
    double *ax = REAL(GET_SLOT(ans, install("x"))), *F, Di, tmp;

    SEXP tmp2 = PROTECT(allocVector(REALSXP, n));
    setAttrib(ans, install("F"), tmp2);

    F = REAL(getAttrib(ans, install("F")));
    for (int i = 0; i < n; i++) {
	int p = sire[i] - 1, q = dam[i] - 1;
	if (sire[i] == NA_INTEGER) {
	    F[i] = 0;
	    Di = (dam[i] == NA_INTEGER) ? 1 : sqrt(0.75 - 0.25 * F[q]);
	} else {
	    if (dam[i] == NA_INTEGER) { /* sire only */
		F[i] = 0;
		Di = sqrt(0.75 - 0.25 * F[p]);
	    } else {		/* both parents in pedigree */
		Di = sqrt(0.5 - 0.25 * (F[p] + F[q]));
		F[i] = NA_REAL;
		if ((ap[i + 1] - ap[i]) > 1) {	  /* skip if no progeny */
		    if (p > q) {j = p; p = q; q = j;} /* ensure p <= q */
		    F[i] = 0;
		    for (int j = 0; j <= p; j++) {
			tmp = 0;
			for (int k = ap[j]; k < ap[j + 1] && ai[k] <= q; k++) {
			    int ii = ai[k];
			    if (ii == p) tmp = ax[k];
			    if (ii == q) F[i] += tmp * ax[k]/2;
			}
		    }
		}
	    }
	}
	for (int j = ap[i]; j < ap[i + 1]; j++) ax[j] *= Di;
    }
    UNPROTECT(1);
    return ans;
}

/**
 * Create the inbreeding coefficients according to the algorithm given
 * in "Comparison of four direct algorithms for computing inbreeding
 * coefficients" by Mehdi Sargolzaei and Hiroaki Iwaisaki, Animal
 * Science Journal (2005) 76, 401--406.  This function is a modified
 * version of the code published in an appendix to that paper.
 *
 * @param x a pedigree object
 *
 * @return a list of the inbreeding coefficients
 */

SEXP pedigree_inbreeding(SEXP x)
{
    SEXP ans, sp = GET_SLOT(x, install("sire"));
    int i, j, t, n = LENGTH(sp), S, D;
    int *SI, *MI,		/* start and minor */
	*sire = INTEGER(sp),
	*dam = INTEGER(GET_SLOT(x, install("dam")));
    double *F = Calloc(n + 1, double), /* inbreeding coefficients */
      *L = Calloc(n + 1, double), *B = Calloc(n + 1, double);
    int *Anc = Calloc(n + 1, int),	/* ancestor */
	*LAP = Calloc(n + 1, int); 	/* longest ancestoral path */
    R_CheckStack();
    
    for (i = 0; i < n; i++) {     /* Replace NA's by zeros */
	if (sire[i] == NA_INTEGER) sire[i] = 0;
	if (dam[i] == NA_INTEGER) dam[i] = 0;
    }
    F[0] =-1; LAP[0] =-1; /* set F and lap for unknown parents */
    for(i = 1, t = -1; i <= n; i++) { 	/* evaluate LAP and its maximum */
	S = sire[i-1]; D = dam[i-1]; /* parents of animal i-1 */
	LAP[i] = ((LAP[S] < LAP[D]) ? LAP[D] : LAP[S]) + 1;
	if (LAP[i] > t) t = LAP[i];
    }
    SI = Calloc(t + 1, int);
    MI = Calloc(t + 1, int);
    for(i = 0; i <= t ; ++i) SI[i] = MI[i] = 0; /* initialize start and minor */
    for(i = 1; i <= n; i++) { 	/* evaluate F */
	S = sire[i-1]; D = dam[i-1]; /* parents of animal i */
	B[i] = 0.5 - 0.25 * (F[S] + F[D]); 
				/* adjust start and minor */
	for (j = 0; j < LAP[i]; j++) {++SI[j]; ++MI[j];} 
	if (S == 0 || D == 0) { /* both parents unknown */
	    F[i] = L[i] = 0; continue;
	}
	if(S == sire[i-2] && D == dam[i-2]) { /* full-sib with last animal */
	    F[i] = F[i-1]; L[i] = L[i-1]; continue;
	}
    
	F[i] = -1; L[i] = 1; 
	t = LAP[i]; /* largest lap group number in the animal's pedigree */
	Anc[MI[t]++] = i; /* initialize Anc and increment MI[t] */
	while(t > -1) { /* from the largest lap group to zero */
	    j = Anc[--MI[t]]; /* next ancestor */
	    S = sire[j-1]; D = dam[j-1]; /* parents of the ancestor */
	    if (S) {
		if (!L[S]) Anc[MI[LAP[S]]++] = S; 
				/* add sire in its lap group in Anc
				 * array if it is not added yet and
				 * increment the minor index for the group */ 
		L[S] += 0.5 * L[j]; /* contribution to sire */
	    }
	    if (D) {
		if (!L[D]) Anc[MI[LAP[D]]++] = D;
		L[D] += 0.5 * L[j]; /* contribution to dam */
	    }
	    F[i] += L[j] * L[j] * B[j];
	    L[j] = 0; /*clear L[j] for the evaluation of the next animal */
	    if (MI[t] == SI[t]) --t; /* move to the next lap group when
				      * all ancestors in group t have been
				      * evaluated */
	} 
    }
    ans = PROTECT(allocVector(REALSXP, n));
    Memcpy(REAL(ans), F + 1, n);

    for (i = 0; i < n; i++) {     /* Restore the NA's */
	if (!sire[i]) sire[i] = NA_INTEGER;
	if (!dam[i]) dam[i] = NA_INTEGER;
    }
    Free(F); Free(L); Free(B); Free(Anc); Free(LAP); Free(SI); Free(MI);
    UNPROTECT(1);
    return ans;
}



R/getASelfing.R
#' Extends the pedigree according to number of selfing cycles 
#' and also optionally computes the Additive Relationship Matrix for that pedigree.
#' @param ID is a vector of individual IDs
#' @param Par1 vector of IDs of one of the parents
#' @param Par2 vector of IDs of the other parent
#' @param nCycles vector that indicates number of selfing cycles for each individual. 
#' @param nCyclesDefault default value of nCycles
#' @param sepChar 
#' @param verbose 
#' @param fileNewPed Output csv file (comma separated value) with columns 'label', 'sire', 'dam', with the full pull pedigree expanded taking into account the selfing cycles
#' @param computeA Indicates if the A matrix is to be computed
#' @return Returns A matrix computed for the extended pedigree if computeA=TRUE
#' @export

#R Code

getASelfing=function(ID,Par1,Par2,nCycles,nCyclesDefault, sepChar='-F', verbose=FALSE, fileNewPed=NULL, computeA=TRUE)
{

## Completing and sorting pedigree
TMP<-as.data.frame(editPed(sire=Par1,dam=Par2,label=ID))
 PED<-merge(x=TMP,y=data.frame(ID=ID,nCycles=nCycles,stringsAsFactors=FALSE), all.x=TRUE,by.x=1,by.y=1,sort=FALSE)
 PED=PED[order(PED$gene),]
 
## Adding number of selfing cycles for lines that miss that information 
 index=is.na(PED$nCycles)
 if(length(index)>0){ PED$nCycles[index]=nCyclesDefault}

## Extending the pedigree by adding selfing
  newPED<-PED[integer(),]

  row<-0

  for(i in 1:nrow(PED)){

		id<-PED$label[i]
		
		generation<-PED$gene[i]
		cycles<-PED$nCycles[i]
		row<-row+1	
		newPED[row,1]<-paste(id,0,sep=sepChar)
		newPED[row,2:3]<-PED[i,2:3]
		newPED[row,4]<-paste(PED[i,4],0,sep=sepChar)
	
		# now we update the id in the sire and dam columns
		newID<-paste(id,cycles,sep=sepChar)
		tmp<-which(PED$sire==id)
		if(length(tmp)>0){
			PED$sire[tmp]<-newID
		}
				
		tmp<-which(PED$dam==id)
		if(length(tmp)>0){
			PED$dam[tmp]<-newID
		}
	
		if(cycles>0){
			for(j in 1:cycles){
				row<-row+1
				newPED[row,1]<-paste(id,j,sep=sepChar)
				newPED[row,2:3]<-rep(newPED[(row-1),1],2)
				newPED[row,4]<-paste(generation,j,sep=sepChar)
			}
		
		}
		if(verbose){print(i)}
	}
		
	if(!is.null(fileNewPed)){
		write.table(newPED,file=fileNewPed,quote=FALSE,row.names=FALSE,col.names=TRUE,sep=",")
	}
	
	if(computeA){
		newPED<-pedigree(sire=newPED$sire,dam=newPED$dam,label=newPED$label)
		A<-getA(newPED)
		tmp<-rownames(A)
		tmp2<-paste(ID,nCycles,sep=sepChar)
		tmpIndex<-which(tmp%in%tmp2)
		A<-A[tmpIndex,tmpIndex]
		TMP<-matrix(unlist(strsplit(rownames(A),split=sepChar)),ncol=2,byrow=TRUE)[,1]
		rownames(A)<-TMP
		colnames(A)<-TMP
		return(A)
	}
}   
   

#This code will be used in EXAMPLES!!!!
if(FALSE)
{

pedigree_data=read.csv(file="~/Desktop/Bayer/MCP2009_pedigree.csv",header=TRUE)


nCycles=substr(pedigree_data$Generation,start=2,stop=2)
nCycles=as.integer(nCycles)
ID=pedigree_data$id
Par1=pedigree_data$Par1
Par2=pedigree_data$Par2
A=getASelfing(ID=ID,Par1=Par1,Par2=Par2,nCycles=nCycles,nCyclesDefault=6)
 
sum(cov2cor(A)[row(A)>col(A)]>.5)
	
A[c('57','62'),c('57','62')]



plot(diag(A),ylim=c(0,2))
 image(A)
 A<-as.matrix(A)

 SVD<-svd(A,nu=2)
 eigenValues<-SVD$d^2
 plot(eigenValues,ylab='Eeigen-value (A)',col=4,cex=.7)
 
 plot(cumsum(eigenValues)/sum(eigenValues),ylab= 'Proportion of variance',col=4,cex=.7,type='o');abline(h=1,lty=2)
 
 plot(SVD$u,col=2,xlab='1st PC',ylab='2nd PC')

}


R/getASelfingV2.R
#R Code

getASelfing=function(ID,Par1,Par2,nCycles,nCyclesDefault, sepChar='-F', verbose=FALSE, fileNewPed=NULL, computeA=TRUE)
{

## Completing and sorting pedigree
TMP<-as.data.frame(editPed(sire=Par1,dam=Par2,label=ID))
 PED<-merge(x=TMP,y=data.frame(ID=ID,nCycles=nCycles,stringsAsFactors=FALSE), all.x=TRUE,by.x=1,by.y=1,sort=FALSE)
 PED=PED[order(PED$gene),]
 
## Adding number of selfing cycles for lines that miss that information 
 index=is.na(PED$nCycles)
 if(length(index)>0){ PED$nCycles[index]=nCyclesDefault}

## Extending the pedigree by adding selfing
  newPED<-PED[integer(),]

  row<-0

  for(i in 1:nrow(PED)){

		id<-PED$label[i]
		
		generation<-PED$gene[i]
		cycles<-PED$nCycles[i]
		row<-row+1	
		newPED[row,1]<-paste(id,0,sep=sepChar)
		newPED[row,2:3]<-PED[i,2:3]
		newPED[row,4]<-paste(PED[i,4],0,sep=sepChar)
	
		if(cycles>1){
			for(j in 1:cycles){
				row<-row+1
				newPED[row,1]<-paste(id,j,sep=sepChar)
				newPED[row,2:3]<-rep(newPED[(row-1),1],2)
				newPED[row,4]<-paste(generation,j,sep=sepChar)
			}
		
			# now we update the id
			newID<-paste(id,cycles,sep=sepChar)
			#tmp<-grep(PED$sire,pattern=id)#*#
			tmp<-which(PED$sire==id)
			if(length(tmp)>0){
				PED$sire[tmp]<-newID
			}
				
			#tmp<-grep(PED$dam,pattern=id)
			tmp<-which(PED$dam==id)
			if(length(tmp)>0){
				PED$dam[tmp]<-newID
			}
		
		}
		if(verbose){print(i)}
	}
		
	if(!is.null(fileNewPed)){
		write.table(newPED,file=fileNewPed,quote=FALSE,row.names=FALSE,col.names=TRUE,sep=",")
	}
	
	if(computeA){
		newPED<-pedigree(sire=newPED$sire,dam=newPED$dam,label=newPED$label)
		A<-getA(newPED)
		tmp<-rownames(A)
		tmp2<-paste(ID,nCycles,sep=sepChar)
		tmpIndex<-which(tmp%in%tmp2)
		A<-A[tmpIndex,tmpIndex]
		TMP<-matrix(unlist(strsplit(rownames(A),split=sepChar)),ncol=2,byrow=TRUE)[,1]
		rownames(A)<-TMP
		colnames(A)<-TMP
		return(A)
	}
}   
   

#This code will be used in EXAMPLES!!!!
if(FALSE)
{

pedigree_data=read.csv(file="~/Desktop/Bayer/MCP2009_pedigree.csv",header=TRUE)


nCycles=substr(pedigree_data$Generation,start=2,stop=2)
nCycles=as.integer(nCycles)
ID=pedigree_data$id
Par1=pedigree_data$Par1
Par2=pedigree_data$Par2
A=getASelfing(ID=ID,Par1=Par1,Par2=Par2,nCycles=nCycles,nCyclesDefault=6)
 
sum(cov2cor(A)[row(A)>col(A)]>.5)
	
A[c('57','62'),c('57','62')]



plot(diag(A),ylim=c(0,2))
 image(A)
 A<-as.matrix(A)

 SVD<-svd(A,nu=2)
 eigenValues<-SVD$d^2
 plot(eigenValues,ylab='Eeigen-value (A)',col=4,cex=.7)
 
 plot(cumsum(eigenValues)/sum(eigenValues),ylab='Proportion of variance',col=4,cex=.7,type='o');abline(h=1,lty=2)
 
 plot(SVD$u,col=2,xlab='1st PC',ylab='2nd PC')

}


R/pedigree.R
#### "pedigree" class methods

#' Constructor for pedigree objects
#'
#' A simple constructor for a pedigree object.  The main point for the
#' constructor is to use coercions to make the calls easier.
#'
#' @param sire integer vector or factor representation of the sires
#' @param dam integer vector or factor representation of the dams
#' @param label character vector of labels
#' @return an pedigree object of class \linkS4class{pedigree}
#' @note \code{sire}, \code{dam} and \code{label} must all have the
#'   same length and all labels in \code{sire} and \code{dam} must occur
#'   in \code{label}
#' @export
#' @examples
#' ped <- pedigree(sire=c(NA,NA,1,1,4,5), dam=c(NA,NA,2,NA,3,2), label=1:6)
pedigree <- function(sire, dam, label) {
    n <- length(sire)
    labelex <- c(label, NA, 0)
    stopifnot(n == length(dam),
              n == length(label),
              all(sire %in% labelex),
              all(dam %in% labelex))
    sire <- as.integer(factor(sire, levels = label))
    dam <- as.integer(factor(dam, levels = label))
    sire[sire < 1 | sire > n] <- NA
    dam[dam < 1 | dam > n] <- NA
    new("pedigree", sire = sire, dam = dam,
        label = as.character(label))
}

setAs("pedigree", "sparseMatrix", # representation as T^{-1}
      function(from) {
	  sire <- from@sire
	  n <- length(sire)
	  animal <- seq_along(sire)
	  j <- c(sire, from@dam)
	  ind <- !is.na(j)
	  as(new("dtTMatrix", i = rep.int(animal, 2)[ind] - 1L,
		 j = j[ind] - 1L, x = rep.int(-0.5, sum(ind)),
		 Dim = c(n,n), Dimnames = list(from@label, NULL),
		 uplo = "L", diag = "U"), "dtCMatrix")
      })

## these data frames are now storage efficient but print less nicely
setAs("pedigree", "data.frame",
      function(from)
      data.frame(sire = from@sire, dam = from@dam,
		 row.names = from@label))

#' Convert a pedigree to a data frame
#'
#' Express a pedigree as a data frame with \code{sire} and
#' \code{dam} stored as factors.  If the pedigree is an object of
#' class pedinbred then the inbreeding coefficients are
#' appended as the variable \code{F}
#'
#' @param x a pedigree object of class \linkS4class{pedigree}
#' @return a data frame
ped2DF <- function(x) {
    stopifnot(is(x, "pedigree"))
    lab <- x@label
    lev <- seq_along(lab)
    ans <- data.frame(sire = factor(x@sire, levels = lev, labels = lab),
                      dam  = factor(x@dam,  levels = lev, labels = lab),
                      row.names = lab)
    if (is(x, "pedinbred")) ans <- cbind(ans, F = x@F)
    ans
}

setMethod("show", signature(object = "pedigree"),
	  function(object) print(ped2DF(object)))

setMethod("head", "pedigree", function(x, ...)
	  do.call("head", list(x = ped2DF(x), ...)))

setMethod("tail", "pedigree", function(x, ...)
	  do.call("tail", list(x = ped2DF(x), ...)))

#' @useDynLib pedigreeTools pedigree_chol
setMethod("chol", "pedigree",
          function(x, pivot, LINPACK) {
              ttrans <- Matrix::solve(Matrix::t(as(x, "dtCMatrix")))
              .Call(pedigree_chol, x,
                    as(.Call("Csparse_diagU2N", Matrix::t(ttrans), PACKAGE = "Matrix"),
                       "dtCMatrix"))
          })

#' Inbreeding coefficients from a pedigree
#'
#' Create the inbreeding coefficients according to the algorithm given
#' in "Comparison of four direct algorithms for computing inbreeding
#' coefficients" by Mehdi Sargolzaei and Hiroaki Iwaisaki, Animal
#' Science Journal (2005) 76, 401--406.
#'
#' @param ped an object that inherits from class \linkS4class{pedigree}
#' @return the inbreeding coefficients as a numeric vector
#' @export
#' @useDynLib pedigreeTools pedigree_inbreeding
#' @examples
#' ped <- pedigree(sire=c(NA,NA,1,1,4,5), dam=c(NA,NA,2,NA,3,2), label=1:6)
#' inbreeding(ped)
inbreeding <- function(ped) {
    stopifnot(is(ped, "pedigree"))
    .Call(pedigree_inbreeding, ped)
}

#' Diagonal of D in the A = TDT' factorization.
#'
#' Determine the diagonal factor in the decomposition of the
#' relationship matrix A as TDT' where T is unit lower triangular.
#'
#' @param ped an object that inherits from class \linkS4class{pedigree}
#' @return a numeric vector
#' @export
#' @examples
#' ped <- pedigree(sire=c(NA,NA,1,1,4,5), dam=c(NA,NA,2,NA,3,2), label=1:6)
#' Dmat(ped)
Dmat <- function(ped)
{
    F <- inbreeding(ped)
    sire <- ped@sire
    dam <- ped@dam
    Fsire <- ifelse(is.na(sire), -1, F[sire])
    Fdam <-  ifelse(is.na(dam), -1, F[dam])
    ans <- 1 - 0.25 * (2 + Fsire + Fdam)
    names(ans) <- ped@label
    ans
}

#' Relationship factor from a pedigree
#'
#' Determine the right Cholesky factor of the relationship matrix for
#' the pedigree \code{ped}, possibly restricted to the specific labels
#' that occur in \code{labs}.
#'
#' @param ped a pedigree that includes the individuals who occur in svec
#' @param labs a character vector or a factor giving the labels to
#'   which to restrict the relationship matrix. If \code{labs} is a
#'   factor then the levels of the factor are used as the labels.
#'   Default is the complete set of labels in the pedigree.
#' @return an object that inherits from \linkS4class{CHMfactor}
#' @export
#' @examples
#' ped <- pedigree(sire=c(NA,NA,1,1,4,5), dam=c(NA,NA,2,NA,3,2), label=1:6)
#' relfactor(ped)
relfactor <- function(ped, labs)
{
    stopifnot(is(ped, "pedigree"))
    if (missing(labs))                  # square case
        return(Matrix::Diagonal(x = sqrt(Dmat(ped))) %*%
               Matrix::solve(Matrix::t(as(ped, "sparseMatrix"))))
    labs <- factor(labs) # drop unused levels from a factor
    stopifnot(all(labs %in% ped@label))
    rect <- Matrix::Diagonal(x = sqrt(Dmat(ped))) %*%
        Matrix::solve(Matrix::t(as(ped, "sparseMatrix")), # rectangular factor
              as(factor(ped@label, levels = ped@label),"sparseMatrix"))
    tmpA<-Matrix::crossprod(rect)
    tmp<- ped@label %in% labs
    tmpA<-tmpA[tmp,tmp]

    orlab <- order(as.numeric(factor(labped<-ped@label[tmp], levels=labs, ordered=T)))
    labped<- as.character(labped[orlab])
    tmpA  <- tmpA[orlab, orlab]
    stopifnot(all.equal(as.character(labped), as.character(labs)))
    relf<-Matrix::chol(tmpA)
    dimnames(relf)[[1]]<- dimnames(relf)[[2]]<-labs
    relf
}

#' Inverse of the Relationship Matrix
#'
#' @param ped a pedigree that includes the individuals who occur in svec
#'   which to restrict the relationship matrix. If \code{labs} is a
#'   factor then the levels of the factor are used as the labels.
#'   Default is the complete set of labels in the pedigree.
#' @return an object that inherits from \linkS4class{CHMfactor}
#' @export
#' @examples
#' ped <- pedigree(sire=c(NA,NA,1,1,4,5), dam=c(NA,NA,2,NA,3,2), label=1:6)
#' getAInv(ped)
getAInv <- function(ped)
{
    stopifnot(is(ped, "pedigree"))
    T_Inv <- as(ped, "sparseMatrix")
    D_Inv <- Matrix::diag(1/Dmat(ped))
    aiMx<-Matrix::t(T_Inv) %*% D_Inv %*% T_Inv
    dimnames(aiMx)[[1]]<-dimnames(aiMx)[[2]] <-ped@label
    aiMx
}

#' Additive Relationship Matrix
#'
#' Returns the additive relationship matrix for the pedigree \code{ped}.
#'
#' @param ped a pedigree that includes the individuals who occur in svec
#'   which to restrict the relationship matrix. If \code{labs} is a
#'   factor then the levels of the factor are used as the labels.
#'   Default is the complete set of labels in the pedigree.
#' @return an object that inherits from \linkS4class{CHMfactor}
#' @keywords array
#' @export
#' @examples
#' ped <- pedigree(sire=c(NA,NA,1,1,4,5), dam=c(NA,NA,2,NA,3,2), label=1:6)
#' getA(ped)
getA <- function(ped)
{
    stopifnot(is(ped, "pedigree"))
    aMx<-Matrix::crossprod(relfactor(ped))
    dimnames(aMx)[[1]]<-dimnames(aMx)[[2]] <-ped@label
    aMx
}

#' Counts number of generations of ancestors for one subject. Use recursion.
#'
#' @param pede data frame with a pedigree and a column for the number of
#'   generations of each subject.
#' @param id subject for which we want the number of generations.
#' @param ngen number of generation
#' @return a data frame object with the pedigree and generation of
#'   ancestors for subject id.
getGenAncestors <- function(pede, id, ngen=NULL){
    j <- which(pede$id==id)
    parents <- c(pede$sire[j], pede$dam[j])
    parents <- parents[!is.na(parents)]
    np <- length(parents)
    if(np==0)
    {
        pede$gene[j] <-0
        return(pede)
    }
    ## get the number of generations in parent number one
    tmpgenP1 <- pede$gene[pede$id==parents[1]]
    if( is.na(tmpgenP1))
    {
        #if ngen is not null, and not cero, ngen<- ngen-1
        #if ngen is cero, do not call recurrsively anymore
        pede <- getGenAncestors(pede, parents[1])
        genP1  <- 1 + pede$gene[pede$id==parents[1]]
    }  else {
        genP1 <- 1 + tmpgenP1
    }
    ## find out if there is a parent number two
    if (np==2){
        tmpgenP2 <- pede$gene[pede$id==parents[2]]
        if( is.na(tmpgenP2))
        {
            pede <- getGenAncestors(pede, parents[2])
            genP2  <- 1 + pede$gene[pede$id==parents[2]]
        }  else {
            genP2 <- 1 + tmpgenP2
        }
        genP1 <- max(genP1, genP2)
    }
    pede$gene[j] <- genP1
    ## print(paste('id:', id, ', gen:', genP1, ', row:', j))
    pede
}

#' Edits a disordered or incomplete pedigree.
#'
#' 1_ add labels for the sires and dams not listed as labels before.
#' 2_ order pedigree based on recursive calls to getGenAncestors.
#'
#' @param sire integer vector or factor representation of the sires
#' @param dam integer vector or factor representation of the dams
#' @param label character vector of labels
#' @param verbose logical to print the row of the pedigree that the
#'   function is ordering. Default is FALSE.
#' @return a data frame with the pedigree ordered.
#' @export
#' @examples
#' pede <- data.frame(sire=as.character(c(NA,NA,NA,NA,NA,1,3,5,6,4,8,1,10,8)),
#'                    dam=as.character(c(NA,NA,NA,NA,NA,2,2,NA,7,7,NA,9,9,13)),
#'                    label=as.character(1:14))
#' pede <- pede[sample(replace=FALSE, 1:14),]
#' pede <- editPed(sire=pede$sire, dam=pede$dam, label=pede$label)
#' ped <- with(pede, pedigree(label=label, sire=sire, dam=dam))
editPed <- function(sire, dam, label, verbose = FALSE)
{
    nped <- length(sire)
    if (nped != length(dam))  stop("sire and dam have to be of the same length")
    if (nped != length(label)) stop("label has to be of the same length than sire and dam")
    tmp <- unique(sort(c(as.character(sire), as.character(dam))))

    missingP <-NULL
    if(any(completeId <- ! tmp %in% as.character(label))) missingP <- tmp[completeId]
    labelOl <- c(as.character(missingP),as.character(label))
    sireOl <- c(rep(NA, times=length(missingP)),as.character(sire))
    damOl  <- c(rep(NA, times=length(missingP)),as.character(dam))
    sire <- as.integer(factor(sireOl, levels = labelOl))
    dam  <- as.integer(factor(damOl, levels = labelOl))
    nped <-length(labelOl)
    label <-1:nped
    sire[!is.na(sire) & (sire<1 | sire>nped)] <- NA
    dam[!is.na(dam) & (dam < 1 | dam > nped)] <- NA
    pede <- data.frame(id= label, sire= sire, dam= dam, gene=rep(NA, times=nped))
    noParents <- (is.na(pede$sire) & is.na(pede$dam))
    pede$gene[noParents] <- 0
    for(i in 1:nped){
        if(verbose) print(i)
        if(is.na(pede$gene[i])){
            id <-pede$id[i]
            pede <-getGenAncestors(pede, id)
        }}
    ord<- order(pede$gene)
    ans<-data.frame(label=labelOl, sire=sireOl, dam=damOl, gene=pede$gene,
                    stringsAsFactors =F)
    ans[ord,]
}

#' Subsets a pedigree for a specified vector of individuals upto a 
#' specified number of previous generations using Recursion.
  
#' @param ped Data Frame pedigree to be subset
#' @param selectVector Vector of individuals to select from pedigree
#' @param ngen Number of previous generations of parents to select starting from selectVector. 
  
#' @return Returns Subsetted pedigree as a DataFrame. 
#' @export


prunePed <- function(ped,selectVector,ngen=2){
  
  ped <- as.matrix(ped)

  returnPed <- matrix(c(NA,NA,NA),nrow=1,ncol=3)
                          
  findBase <- ped[,"label"] %in% selectVector 
  basePed <- ped[findBase,]
  findSire <- ped[,"label"] %in% basePed[,"sire"]
  findDam <- ped[,"label"] %in% basePed[,"dam"]
  
  newSelVec <- ped[findSire|findDam,"label"]
  newSelVec <- newSelVec[!(newSelVec %in% selectVector)]
  
  if(ngen!=-1){
  returnPed <- basePed
  returnPed <- unique(rbind(returnPed,prunePed(ped,newSelVec,ngen-1)))
  returnPed <- returnPed[rowSums(is.na(returnPed))!=3,]
  }else{return(returnPed)}

return(as.data.frame(returnPed))

}


